### PR TITLE
[plugin-sauce-connect] Implement sauce connect

### PR DESCRIPTION
**Pull request by ananyevN**

### CODE REVIEW

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/AbstractVividusWebDriverFactory.java [hunk 1]
```diff
@@ -45,10 +45,13 @@ public AbstractVividusWebDriverFactory(IBddRunContext bddRunContext,
     public VividusWebDriver create()
     {
         VividusWebDriver vividusWebDriver = createVividusWebDriver(bddRunContext.getRunningStory());
+        configureVividusWebDriverWithProxy(vividusWebDriver);
         configureVividusWebDriver(vividusWebDriver);
         return vividusWebDriver;
     }
 
+    protected abstract void configureVividusWebDriverWithProxy(VividusWebDriver vividusWebDriver);
```
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-extension-selenium/src/main/java/org/vividus/selenium/AbstractVividusWebDriverFactory.java:53]: We shouldn't set PROXY desiredCapability if sauce connect is on.
So I have created 2 implementations configureVividusWebDriverWithProxy

File: vividus-plugin-sauce-labs/build.gradle [hunk 2]
```diff
@@ -0,0 +1,14 @@
+project.description = 'Vividus plugin for Sauce Labs'
+
+dependencies {
+    implementation project(':vividus-plugin-web-app')
+    implementation project(':vividus-bdd-engine')
+    implementation project(':vividus-extension-selenium')
+
+    implementation(group: 'javax.inject', name: 'javax.inject', version: versions.javaxInject)
```
Comment (ikalinin1) [inline vividus-plugin-sauce-labs/build.gradle:8]: could you please use constructor based injections

File: vividus-plugin-web-app/src/main/resources/properties/profile/web/profile.properties [hunk 3]
```diff
@@ -15,12 +15,19 @@ selenium.grid.url=https://user:accessKey@ondemand.saucelabs.com/wd/hub
 selenium.grid.capabilities.browserName=
 selenium.grid.capabilities.version=
 
+selenium.grid.username=
+selenium.grid.password=
+
+sauceLabs.enabled=false
```
Comment (ikalinin1) [inline vividus-plugin-web-app/src/main/resources/properties/profile/web/profile.properties:21]: @valfirst I propose to add property useLatestVersion from the very start :) It could be very useful

**Review by ikalinin1 — CHANGES_REQUESTED** [review]

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/ControllingMetaTag.java [hunk 4]
```diff
@@ -61,6 +61,18 @@ protected void setCapability(DesiredCapabilities capabilities, MetaWrapper meta)
                 .forEach((k, v) -> capabilities.setCapability(StringUtils.substringAfter(k, getMetaTagName()),
                     adjustType(v)));
         }
+    },
+    SAUCE_CONNECT("sauceConnect")
```
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-extension-selenium/src/main/java/org/vividus/selenium/ControllingMetaTag.java:65]: done

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/SauceLabsCapabilityType.java [hunk 5]
```diff
@@ -28,4 +28,5 @@
     String DEVICE = "device";
     String CAPABILITIES = "capabilities";
     String SELENIUM_VERSION = "seleniumVersion";
+    String TUNNEL_IDENTIFIER = "tunnelIdentifier";
```
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-extension-selenium/src/main/java/org/vividus/selenium/SauceLabsCapabilityType.java:31]: done

File: vividus-plugin-sauce-labs/src/main/java/org/vividus/sauce/ConfigurationChecker.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+ * limitations under the License.
+ */
+
+package org.vividus.sauce;
+
+import org.vividus.ui.web.configuration.WebApplicationConfiguration;
+
+import javax.inject.Inject;
+
+public class ConfigurationChecker
```
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-plugin-sauce-labs/src/main/java/org/vividus/sauce/ConfigurationChecker.java]: done

File: vividus-plugin-sauce-labs/src/main/java/org/vividus/sauce/SynchronizedUnpackingSauceConnectTunnelManager.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+package org.vividus.sauce;
+
+import java.io.File;
+import java.io.IOException;
+
+import com.saucelabs.ci.sauceconnect.SauceConnectFourManager;
+
+@SuppressWarnings("checkstyle:ImportOrder")
```
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-plugin-sauce-labs/src/main/java/org/vividus/sauce/SynchronizedUnpackingSauceConnectTunnelManager.java:24]: Thanks, removed

File: vividus-plugin-sauce-labs/src/main/resources/profile.properties [hunk 8]
```diff
@@ -0,0 +1,2 @@
+sauceLabs.enabled=false
```
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-plugin-sauce-labs/src/main/resources/profile.properties:1]: done

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/ProxyAgentStoryReporter.java [hunk 9]
```diff
@@ -100,9 +100,18 @@ public void afterStory(boolean givenStory)
         }
     }
 
-    private void stopProxy()
+    @Override
+    public void startProxy()
+    {
+        proxy.start();
+        subscribers.forEach(IProxyListener::startProxyNotify);
```
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/ProxyAgentStoryReporter.java]: Thanks, done

File: vividus-plugin-web-app/src/main/java/org/vividus/ui/web/configuration/AuthenticationMode.java [hunk 10]
```diff
@@ -29,8 +30,18 @@ protected String getUrlImpl(String url, String userInfo)
         {
             return url;
         }
+    },
+    SAUCE_CONNECT
```
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-plugin-web-app/src/main/java/org/vividus/ui/web/configuration/AuthenticationMode.java:34]: Moved

File: vividus-plugin-web-app/src/main/resources/properties/profile/web/profile.properties [hunk 11]
```diff
@@ -15,12 +15,19 @@ selenium.grid.url=https://user:accessKey@ondemand.saucelabs.com/wd/hub
...
+sauceLabs.sauce-connect.enabled=false
```
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-plugin-web-app/src/main/resources/properties/profile/web/profile.properties:22]: done

File: vividus-plugin-web-app/src/main/resources/properties/profile/web/profile.properties [hunk 12]
```diff
@@ -15,12 +15,19 @@ selenium.grid.url=https://user:accessKey@ondemand.saucelabs.com/wd/hub
...
+
 proxy.enabled=false
 proxy.recording.enabled=false
 proxy.trust-all-servers=false
 proxy.host=localhost
 proxy.ports=0
 proxy.capture-types=REQUEST_CONTENT, RESPONSE_CONTENT, REQUEST_BINARY_CONTENT, RESPONSE_BINARY_CONTENT, REQUEST_HEADERS, RESPONSE_HEADERS
+proxy.noSslBumpDomains=all
```
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-plugin-web-app/src/main/resources/properties/profile/web/profile.properties:30]: removed

File: settings.gradle [hunk 13]
```diff
@@ -30,5 +30,6 @@ include 'vividus-plugin-visual'
 include 'vividus-plugin-web-app-to-rest-api'
 include 'vividus-plugin-web-app'
 include 'vividus-plugin-xml'
+include 'vividus-plugin-sauce-labs'
```
[the remark this answers is not in the document]
Comment (ananyevN) [reply settings.gradle:33]: done

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/AbstractVividusWebDriverFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-extension-selenium/src/main/java/org/vividus/selenium/AbstractVividusWebDriverFactory.java:53]: Because we should have 2 implementations

File: vividus-plugin-sauce-labs/src/main/java/org/vividus/sauce/SauceConnectListener.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+                    sauceConnectOptions.setPort(DEFAULT_HTTPS_PORT);
+                }
+                sauceConnectOptions.setBasicAuthUser(webApplicationConfiguration.getBasicAuthUser());
+            }
+            if (proxy.isStarted())
+            {
+                sauceConnectOptions.setProxy(proxyConfigurator.createSeleniumProxy(false).getHttpProxy());
+            }
+            sauceConnectOptions.setNoSslBumpDomains(noSslBumpDomains);
+            sauceConnectManager.start(sauceConnectOptions);
```
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-plugin-sauce-labs/src/main/java/org/vividus/sauce/SauceConnectListener.java:87]: remove unused if. Thanks

File: vividus-plugin-sauce-labs/build.gradle (hunk 2 above)
Comment (ananyevN) [reply vividus-plugin-sauce-labs/build.gradle:8]: done

File: vividus-plugin-web-app/src/main/resources/properties/profile/web/profile.properties (hunk 3 above)
Comment (ananyevN) [reply vividus-plugin-web-app/src/main/resources/properties/profile/web/profile.properties:21]: Thanks, done

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/AbstractVividusWebDriverFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ananyevN) [reply vividus-extension-selenium/src/main/java/org/vividus/selenium/AbstractVividusWebDriverFactory.java]: need an additional method for local proxy configuring(for non-saucelabs run)

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/ControllingMetaTag.java [hunk 15]
```diff
@@ -104,4 +104,9 @@ public static void setDesiredCapabilitiesFromMeta(DesiredCapabilities capabiliti
             controllingMetaTag.setCapability(capabilities, meta);
         }
     }
+
+    public static boolean isContainedIn(Meta meta, String metaName)
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/ControllingMetaTag.java:108]: no need to make this method a part of ControllingMetaTag

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/WebDriverProvider.java [hunk 16]
```diff
@@ -29,9 +29,9 @@
 
 public class WebDriverProvider implements IWebDriverProvider
 {
-    private IVividusWebDriverFactory vividusWebDriverFactory;
-    private final ConcurrentLinkedQueue<WebDriver> webDrivers = new ConcurrentLinkedQueue<>();
+    @Inject private IVividusWebDriverFactory vividusWebDriverFactory;
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/WebDriverProvider.java:32]: what's the reason of this change?

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/WebDriverProvider.java [hunk 17]
```diff
@@ -29,9 +29,9 @@
...
     @Inject private EventBus eventBus;
+    private final ConcurrentLinkedQueue<WebDriver> webDrivers = new ConcurrentLinkedQueue<>();
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/WebDriverProvider.java:34]: noisy changes

File: vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsOptionName.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.sauce;
+
+public enum SauceLabsOptionName
```
Comment (valfirst) [inline vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsOptionName.java:19]: is it an option or capability?

File: vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsWebDriverFactory.java [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+     *
+     * @param desiredCapabilities - driver's desired capabilities
+     * @param key                 - sauce-option key
+     * @param value               - sauce-option value
+     */
+    @SuppressWarnings("unchecked")
+    private void addSauceOption(DesiredCapabilities desiredCapabilities,
+            SauceLabsOptionName key, Object value)
+    {
+        // @formatter:off
```
Comment (valfirst) [inline vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsWebDriverFactory.java:85]: why is it needed?

File: vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsWebDriverFactory.java [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+                sauceOptions.put(key.getOptionName(), value),
+                () -> desiredCapabilities.setCapability(SAUCE_OPTION, adjustNewSauceOptionsCapability(key, value)));
+        // @formatter:on
+    }
+
+    private Map<String, Object> adjustNewSauceOptionsCapability(SauceLabsOptionName key, Object value)
+    {
+        Map<String, Object> sauceOptions = new HashMap<>();
+        sauceOptions.put(key.getOptionName(), value);
+        return sauceOptions;
```
Comment (valfirst) [inline vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsWebDriverFactory.java:97]: `Map.of(...)`

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/proxy/ProxyEvent.java [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.bdd.proxy;
+
+import org.vividus.proxy.IProxy;
+
+public class ProxyEvent
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/proxy/ProxyEvent.java:21]: it's better to have 2 events: `ProxyStartedEvent` and `ProxyStoppedEvent`

File: vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsWebDriverFactory.java [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+
+    public SauceLabsWebDriverFactory(IWebDriverFactory webDriverFactory, IBddRunContext bddRunContext,
+            IWebDriverManagerContext webDriverManagerContext, IProxy proxy,
+            IBrowserWindowSizeProvider browserWindowSizeProvider, IWebDriverManager webDriverManager)
+    {
+        super(webDriverFactory, bddRunContext, webDriverManagerContext, proxy, browserWindowSizeProvider,
+                webDriverManager);
+    }
+
+    @Override protected void configureProxy(DesiredCapabilities desiredCapabilities)
```
Comment (valfirst) [inline vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsWebDriverFactory.java:51]: why is it empty? and why should it be overridden ?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/ControllingMetaTag.java (hunk 15 above)
Comment (ananyevN) [reply vividus-extension-selenium/src/main/java/org/vividus/selenium/ControllingMetaTag.java:108]: moved

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/WebDriverProvider.java (hunk 17 above)
Comment (ananyevN) [reply vividus-extension-selenium/src/main/java/org/vividus/selenium/WebDriverProvider.java:34]: done

File: vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsOptionName.java (hunk 18 above)
Comment (ananyevN) [reply vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsOptionName.java:19]: done

File: vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsWebDriverFactory.java (hunk 19 above)
Comment (ananyevN) [reply vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsWebDriverFactory.java:85]: just for pretty

File: vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsWebDriverFactory.java (hunk 20 above)
Comment (ananyevN) [reply vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsWebDriverFactory.java:97]: I need a modifiable map for SauceOptions

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/proxy/ProxyEvent.java (hunk 21 above)
Comment (ananyevN) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/proxy/ProxyEvent.java:21]: Thanks,done

File: vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsWebDriverFactory.java (hunk 22 above)
Comment (ananyevN) [reply vividus-plugin-saucelabs/src/main/java/org/vividus/sauce/SauceLabsWebDriverFactory.java:51]: because I don't need to configure proxy while sauce labs run. all that I needed - is configuring with sauce connect

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/WebDriverProvider.java (hunk 16 above)
Comment (ananyevN) [reply vividus-extension-selenium/src/main/java/org/vividus/selenium/WebDriverProvider.java:32]: for non-specifiing a concrete implementation

**Review by valfirst — CHANGES_REQUESTED** [review]

**Comment by valfirst** [comment]
Superseded by #954
